File: handy-js/quiz/q7_parking_lot_spot_allocation_medium.js
/**
 * Q7) Parking Lot Spot Allocation
 *
 * AI-BEST:
 * Interval-based free-space tracking (concept used in scalable implementations).
 * In production, a balanced tree can drive near O(log n) updates/lookups.
 *
 * AI-EASY:
 * Linear scan over spot array for a contiguous free segment.
 * Time: O(n) park in worst case, Space: O(n)
 */

class Vehicle {
  constructor(licensePlate) {
    this.licensePlate = licensePlate;
  }
}

class Motorcycle extends Vehicle {
  getRequiredSpots() {
    return 1;
  }
}

class Car extends Vehicle {
  getRequiredSpots() {
    return 1;
  }
}

class Truck extends Vehicle {
  getRequiredSpots() {
    return 3;
  }
}

class ParkingLotEasy {
  constructor(totalSpots) {
    this.spots = new Array(totalSpots).fill("0");
  }

  park(vehicle) {
    const need = vehicle.getRequiredSpots();
    for (let i = 0; i <= this.spots.length - need; i++) {
      const segment = this.spots.slice(i, i + need);
      if (segment.every((spot) => spot === "0")) {
        for (let j = i; j < i + need; j++) {
          this.spots[j] = vehicle.licensePlate;
        }
        console.log(`[EASY] Parked ${vehicle.constructor.name} ${vehicle.licensePlate} at ${i}-${i + need - 1}`);
        return true;
      }
    }
    console.log(`[EASY] No room for ${vehicle.licensePlate}`);
    return false;
  }

  unpark(licensePlate) {
    let removed = false;
    for (let i = 0; i < this.spots.length; i++) {
      if (this.spots[i] === licensePlate) {
        this.spots[i] = "0";
        removed = true;
      }
    }
    console.log(`[EASY] Unpark ${licensePlate}: ${removed}`);
    return removed;
  }
}

class ParkingLotBest {
  constructor(totalSpots) {
    this.totalSpots = totalSpots;

    // Intervals are inclusive ranges [start, end] of free spots.
    this.freeIntervals = [{ start: 0, end: totalSpots - 1 }];
    this.parked = new Map(); // plate -> { start, size }
  }

  park(vehicle) {
    const need = vehicle.getRequiredSpots();

    for (let i = 0; i < this.freeIntervals.length; i++) {
      const interval = this.freeIntervals[i];
      const available = interval.end - interval.start + 1;
      if (available >= need) {
        const start = interval.start;
        const end = start + need - 1;

        this.parked.set(vehicle.licensePlate, { start, size: need });

        // Consume from interval start; remove interval if fully used.
        if (end === interval.end) {
          this.freeIntervals.splice(i, 1);
        } else {
          interval.start = end + 1;
        }

        console.log(`[BEST] Parked ${vehicle.constructor.name} ${vehicle.licensePlate} at ${start}-${end}`);
        return true;
      }
    }

    console.log(`[BEST] No room for ${vehicle.licensePlate}`);
    return false;
  }

  unpark(licensePlate) {
    const record = this.parked.get(licensePlate);
    if (!record) {
      console.log(`[BEST] Unpark ${licensePlate}: false`);
      return false;
    }

    this.parked.delete(licensePlate);
    const interval = { start: record.start, end: record.start + record.size - 1 };

    // Insert back in sorted order, then merge adjacent/overlapping intervals.
    this.freeIntervals.push(interval);
    this.freeIntervals.sort((a, b) => a.start - b.start);

    const merged = [];
    for (const curr of this.freeIntervals) {
      if (merged.length === 0 || merged[merged.length - 1].end + 1 < curr.start) {
        merged.push({ ...curr });
      } else {
        merged[merged.length - 1].end = Math.max(merged[merged.length - 1].end, curr.end);
      }
    }
    this.freeIntervals = merged;

    console.log(`[BEST] Unpark ${licensePlate}: true`);
    return true;
  }

  debugFreeIntervals() {
    return this.freeIntervals.map((x) => `[${x.start}-${x.end}]`).join(", ");
  }
}

function runDemo() {
  console.log("Q7: Parking Lot Spot Allocation");
  const easyLot = new ParkingLotEasy(8);
  const bestLot = new ParkingLotBest(8);
  const vehicles = [
    new Car("ABC-123"),
    new Truck("BIG-TRK"),
    new Motorcycle("M-42"),
    new Car("CAR-999"),
  ];

  console.log("---- EASY ----");
  for (const v of vehicles) easyLot.park(v);
  easyLot.unpark("BIG-TRK");
  easyLot.park(new Truck("TRK-2"));
  console.log("[EASY] Spots:", easyLot.spots.join(" | "));

  console.log("---- BEST ----");
  for (const v of vehicles) bestLot.park(v);
  console.log("[BEST] Free:", bestLot.debugFreeIntervals());
  bestLot.unpark("BIG-TRK");
  console.log("[BEST] Free after unpark:", bestLot.debugFreeIntervals());
  bestLot.park(new Truck("TRK-2"));
  console.log("[BEST] Free final:", bestLot.debugFreeIntervals());
}

runDemo();

File: handy-python/quiz/q7_parking_lot_spot_allocation_medium.py
"""
Q7) Parking Lot Spot Allocation

AI-BEST:
- Track free contiguous intervals and merge/split on unpark/park.
- In scalable systems this can reach near O(log n) with tree structures.

AI-EASY:
- Linear scan over spot array for contiguous free block.
- Time: O(n) park, Space: O(n)
"""

from __future__ import annotations
from abc import ABC, abstractmethod


class Vehicle(ABC):
    def __init__(self, license_plate: str):
        self.license_plate = license_plate

    @abstractmethod
    def get_required_spots(self) -> int:
        raise NotImplementedError


class Motorcycle(Vehicle):
    def get_required_spots(self) -> int:
        return 1


class Car(Vehicle):
    def get_required_spots(self) -> int:
        return 1


class Truck(Vehicle):
    def get_required_spots(self) -> int:
        return 3


class ParkingLotEasy:
    def __init__(self, total_spots: int):
        self.spots = ["0"] * total_spots

    def park(self, vehicle: Vehicle) -> bool:
        need = vehicle.get_required_spots()
        for i in range(0, len(self.spots) - need + 1):
            if all(spot == "0" for spot in self.spots[i:i + need]):
                for j in range(i, i + need):
                    self.spots[j] = vehicle.license_plate
                print(f"[EASY] Parked {type(vehicle).__name__} {vehicle.license_plate} at {i}-{i + need - 1}")
                return True
        print(f"[EASY] No room for {vehicle.license_plate}")
        return False

    def unpark(self, license_plate: str) -> bool:
        removed = False
        for i in range(len(self.spots)):
            if self.spots[i] == license_plate:
                self.spots[i] = "0"
                removed = True
        print(f"[EASY] Unpark {license_plate}: {removed}")
        return removed


class ParkingLotBest:
    def __init__(self, total_spots: int):
        # free intervals as inclusive pairs [start, end]
        self.free_intervals: list[list[int]] = [[0, total_spots - 1]]
        self.parked: dict[str, tuple[int, int]] = {}

    def park(self, vehicle: Vehicle) -> bool:
        need = vehicle.get_required_spots()
        for idx, interval in enumerate(self.free_intervals):
            start, end = interval
            available = end - start + 1
            if available >= need:
                alloc_start = start
                alloc_end = start + need - 1
                self.parked[vehicle.license_plate] = (alloc_start, need)

                # consume from interval start
                if alloc_end == end:
                    self.free_intervals.pop(idx)
                else:
                    self.free_intervals[idx][0] = alloc_end + 1

                print(f"[BEST] Parked {type(vehicle).__name__} {vehicle.license_plate} at {alloc_start}-{alloc_end}")
                return True

        print(f"[BEST] No room for {vehicle.license_plate}")
        return False

    def unpark(self, license_plate: str) -> bool:
        if license_plate not in self.parked:
            print(f"[BEST] Unpark {license_plate}: False")
            return False

        start, size = self.parked.pop(license_plate)
        self.free_intervals.append([start, start + size - 1])
        self.free_intervals.sort(key=lambda x: x[0])

        # merge touching/overlapping intervals
        merged: list[list[int]] = []
        for current in self.free_intervals:
            if not merged or merged[-1][1] + 1 < current[0]:
                merged.append(current[:])
            else:
                merged[-1][1] = max(merged[-1][1], current[1])
        self.free_intervals = merged

        print(f"[BEST] Unpark {license_plate}: True")
        return True

    def debug_free(self) -> str:
        return ", ".join(f"[{s}-{e}]" for s, e in self.free_intervals)


def run_demo() -> None:
    print("Q7: Parking Lot Spot Allocation")
    vehicles = [Car("ABC-123"), Truck("BIG-TRK"), Motorcycle("M-42"), Car("CAR-999")]

    print("---- EASY ----")
    easy_lot = ParkingLotEasy(8)
    for v in vehicles:
        easy_lot.park(v)
    easy_lot.unpark("BIG-TRK")
    easy_lot.park(Truck("TRK-2"))
    print("[EASY] Spots:", " | ".join(easy_lot.spots))

    print("---- BEST ----")
    best_lot = ParkingLotBest(8)
    for v in vehicles:
        best_lot.park(v)
    print("[BEST] Free:", best_lot.debug_free())
    best_lot.unpark("BIG-TRK")
    print("[BEST] Free after unpark:", best_lot.debug_free())
    best_lot.park(Truck("TRK-2"))
    print("[BEST] Free final:", best_lot.debug_free())


if __name__ == "__main__":
    run_demo()
